File: frontend/src/data/broadcaster.ts
import { TwitchUser } from './twitch-user';
import { UserRoles } from './user-roles';

export interface Broadcaster extends TwitchUser {
  color: string | null;
  roles: UserRoles;
}

const CHANNEL_POINT_TYPES: ReadonlySet<string> = new Set(['affiliate', 'partner']);

export function hasChannelPoints(user: TwitchUser | null): boolean {
  return user !== null && CHANNEL_POINT_TYPES.has(user.broadcasterType.toLowerCase());
}

File: backend/YEPPDash.Api/Data/Twitch/Broadcaster.cs
namespace YEPPDash.Api.Data.Twitch;

public sealed record Broadcaster : TwitchUser
{
    public static Broadcaster From(TwitchUser user, string? color)
    {
        return new Broadcaster
        {
            Id = user.Id,
            Login = user.Login,
            DisplayName = user.DisplayName,
            Type = user.Type,
            BroadcasterType = user.BroadcasterType,
            Description = user.Description,
            ProfileImageUrl = user.ProfileImageUrl,
            OfflineImageUrl = user.OfflineImageUrl,
            CreatedAt = user.CreatedAt,
            Email = user.Email,
            Color = color,
            Roles = new TwitchUserRoles(
                Broadcaster: true,
                Moderator: false,
                Vip: false,
                Editor: false,
                Verified: user.BroadcasterType.Equals("partner", StringComparison.OrdinalIgnoreCase))
        };
    }
}